api/cds.proto

```diff
@@ -0,0 +1,233 @@
+    TcpHealthCheck tcp_health_check = 9;
+  }
+}
+
+// Circuit breaking settings can be specified individually for each defined
+// priority.
+message CircuitBreakers {
+  message Thresholds {
+    // The maximum number of connections that Envoy will make to the upstream
+    // cluster. If not specified, the default is 1024. See the circuit
+    // breaking overview for more information.
+    google.protobuf.UInt32Value max_connections = 1;
+    // The maximum number of pending requests that Envoy will allow to the
+    // upstream cluster. If not specified, the default is 1024. See the circuit
+    // breaking overview for more information.
+    google.protobuf.UInt32Value max_pending_requests = 2;
+    // The maximum number of parallel requests that Envoy will make to the
+    // upstream cluster. If not specified, the default is 1024. See the circuit
+    // breaking overview for more information.
+    google.protobuf.UInt32Value max_requests = 3;
+    // The maximum number of parallel retries that Envoy will allow to the
+    // upstream cluster. If not specified, the default is 3. See the circuit
+    // breaking overview for more information.
+    google.protobuf.UInt32Value max_retries = 4;
+  }
+
+  Thresholds default_priority = 1;
+  Thresholds high_priority = 2;
+}
+
+
+message TcpProtocolOptions {
+}
+
+message Http1ProtocolOptions {
+}
+
+message Http2ProtocolOptions {
+  // Default is false.
+  bool disable_dynamic_table = 1;
+  uint32 per_stream_buffer_limit_bytes = 2;
+}
+
+message GrpcProtocolOptions {
+  uint32 per_stream_buffer_limit_bytes = 1;
+}
+
+message Cluster {
+  // Supplies the name of the cluster which must be unique across all clusters.
+  // The cluster name is used when emitting statistics. The cluster name can be
+  // at most 60 characters long, and must not contain :.
+  google.protobuf.StringValue name = 1;
+
+  // The service discovery type to use for resolving the cluster.
+  enum DiscoveryType {
+    STATIC = 0;
+    STRICT_DNS = 1;
+    LOGICAL_DNS = 2;
+    EDS = 3;
+  }
+  DiscoveryType type = 2;
+
+  // The timeout for new network connections to hosts in the cluster.
+  Duration connect_timeout = 3;
+  // Soft limit on size of the cluster’s connections read and write buffers. If
+  // unspecified, an implementation defined default is applied (1MiB).
+  google.protobuf.UInt32Value per_connection_buffer_limit_bytes = 4;
+
+  // The load balancer type to use when picking a host in the cluster.
+  enum LbPolicy {
+    ROUND_ROBIN = 0;
+    LEAST_REQUEST = 1;
+    RING_HASH = 2;
+    RANDOM = 3;
+  }
+  LbPolicy lb_policy = 5;
+
+  // If the service discovery type is static, strict_dns, or logical_dns
+  // the hosts array is required.
+  repeated Address hosts = 6;
+
+  // Optional active health checking configuration for the cluster. If no
+  // configuration is specified no health checking will be done and all cluster
+  // members will be considered healthy at all times.
+  HealthCheck health_check = 7;
```

Comment:
I wonder whether we should make HealthCheck repeated? This would allow to healthcheck the same endpoint with multiple protocols and produce AND result